Add more Jenkins Books to books page

kanhadharmikk opened the pull request:
Fixes #6019. 

I have added four more Jenkins Books to the books page.

krisstern commented:
Would be a good practice to indicate which issue you are trying to fix with the syntax `Fixes #6019`. I have made the change for this PR.

kanhadharmikk commented:
@krisstern Thank you for your guidance. As it was my first PR I was not aware of it. I will make sure this mistake will not take place next time.

MarkEWaite commented on content/_data/books/Mastering_Jenkins.yml:
```diff
@@ -0,0 +1,19 @@
+title: "Mastering Jenkins"
+url: https://www.amazon.in/Mastering-Jenkins-Jonathan-McAllister/dp/1784390895/ref=sr_1_6?crid=3CUHJHQY7X09V&keywords=jenkins+books&qid=1676393236&sprefix=jenkins+books%2Caps%2C241&sr=8-6
```
I think it is better to link to the worldwide Amazon site and rely on their worldwide site to redirect the user to their local site.
```suggestion
url: https://www.amazon.com/Mastering-Jenkins-Jonathan-McAllister/dp/1784390895
```

kanhadharmikk replied:
@MarkEWaite Thanks for the guidance. I will make the necessary changes and will update my PR.

MarkEWaite commented on content/_data/books/Mastering_Jenkins.yml:
```diff
@@ -0,0 +1,19 @@
+url: https://www.amazon.in/Mastering-Jenkins-Jonathan-McAllister/dp/1784390895/ref=sr_1_6?crid=3CUHJHQY7X09V&keywords=jenkins+books&qid=1676393236&sprefix=jenkins+books%2Caps%2C241&sr=8-6
+author: "Jonathan McAllister"
+publication_year: 2015
+publisher: "Ingram short title"
```
As far as I can tell, this book is published by Packt Publishing rather than Ingram
```suggestion
publisher: "Packt Publishing"
```

MarkEWaite reviewed (changes requested):
Two items I would like to have changed and a request that we have the other URLs updated to use amazon.com

kanhadharmikk commented:
Fixes changes requested #6019 
@MarkEWaite I have made the changes that have been requested.

zbynek commented:
The books page is not really optimized for long descriptions -- formatting gets lost. Maybe instead of copying long texts  from Amazon we could keep the descriptions here short (1 paragraph without formatting) and let the interested reader get more info from the linked Amazon site. 

Alternatively, to support long descriptions we could switch the book format from yaml to AsciiDoc (like other stuff in the `_data` dir).

@MarkEWaite @halkeye thoughts?

halkeye commented:
I'm not sure I'm comfortable hosting images that are not ours, what is the copyright on book images.

Also who vetted the books to know they are legit? We've had issues in the past where retweets have given bad advice.

I do agree that the description isn't processed at all, so it looks pretty bad. They probably can be renamed to .adoc files and use frontmatter without much issue.

gio-neo commented:
I'm new here, having said that here are my thoughts.  On the book images, we (Jenkins) are not profiting from them so I see no concern there.  On vetting the books, I agree and specifically chose to submit a pull request only from an established publisher.  Also, I do think the pull reviewer/editor should reject submittals to the books page that are not beneficial to the intent of the page.  On the format of the page, I was following my own best practice which is to follow the established format that was already there.  I do agree someone could go in and remove the descriptions if they are unnecessary.  Anyway, just wanted to give my feedback, I hope to be able to start contributing more, this was just my first attempt.  Thanks!

kanhadharmikk commented:
@halkeye @zbynek what should I suppose to do next? As this is my first open-source contribution. What should I do next to solve the issue more efficiently?

halkeye commented:
I think just be a little patient. Open source works when people are free, and timezones are hard. 

It could very much be good enough to merge, and we do a followup.

As an idea, and a seperate PR (easier to review smaller ones that have a single purpose), you could look into making the "content/books.html.haml" thingie into a template, then have each book as an .adoc/.md file, and the description as a body. It might look nicer where the description can be full adoc/markdown.

kmartens27 commented:
Hi @kanhadharmikk, thanks for this submission! I was reading over things, and it looks as though the Mastering Jenkins and Learning Continuous Integration with Jenkins entries are twice as large as the other items. The books that we highlight should include _some_ summary of what is contained, but those are far more detailed than necessary. The formatting and structure of the amazon book review does not translate properly to the Jenkins site, and the content itself needs much revising.

kanhadharmikk commented:
Will work on it and make it according to the requirement.

kanhadharmikk commented:
Changes update #6019 
The description of the books Mastering Jenkins and Learning Continous Integration with Jenkins has been made shorter.

kanhadharmikk commented:
What's the status of my PR? @krisstern can you help me with this?

kmartens27 commented:
@kanhadharmikk would it be possible to get higher resolution images? The ones uploaded are less than ideal quality.

kanhadharmikk commented:
Yes, I will upload the higher resolution images of the books.

kmartens27 commented:
Thanks very much @kanhadharmikk, be sure to run them through an image compressor such as compressor.io or something similar. That allows us to keep image quality but decrease the amount of storage it takes.

kanhadharmikk commented:
Thank you so much @kmartens27 for your guidance, I have uploaded the high resolution images and also run through image compressor to keep image quality and decrease the amount of storage it takes.

kmartens27 commented:
Thanks for that @kanhadharmikk, the images are terrific and I appreciate it! Everything looks more aligned and directs the reader to interact with the book page.

There's a bit of a bigger discussion around what book references should look like and include for Jenkins, since we want to provide resources without necessarily recommending one over another.

kanhadharmikk commented:
Removed the bullet points so formatting should look good and added a new Jenkins-Essential book image as it was not rendering in the deployment preview which is of less size compared to the previous one.

kmartens27 commented on content/_data/books/Hands-on_Pipeline_as_YAML_with_Jenkins.yml:
```diff
@@ -0,0 +1,9 @@
+type: Paid
+image: books/Hands-on-pipeline-as-yaml.jpg
+description: |-
+  This book brings solid practical knowledge on how to create YAML pipelines using Jenkins for efficient and scalable CI/CD pipelines. It covers an introduction to various essential topics such as DevOps, DevOps History, Benefits of DevOps Culture, DevOps and Value Streams, DevOps Practices, different types of pipelines such as Build Pipeline, Scripted Pipeline, Declarative Pipeline, YAML Pipelin ...
```
```suggestion
This book provides practical knowledge on how to create YAML pipelines, using Jenkins for efficient and scalable CI/CD pipelines.
There are introductions to various essential topics such as DevOps' history,  culture, and practices. 
This book also reviews different types of pipelines such as Build Pipeline, Scripted Pipeline, Declarative Pipeline, YAML Pipelines, and Blue Ocean.
This book provides a journey for readers to create YAML pipelines for various application systems, including Android, AngularJS, Flutter, and Ionic Cordova.
By the end of this book, you will be able to try out some of the best practices to implement DevOps using Jenkins and YAML.
```
I updated the formatting here to be a sentence per line. This helps with reviewers reading over/providing suggestions as it is easier to go line by line.

I removed one of the sentences in favor of keeping it simple/more reserved. 

I also changed some of the wording to better fit syntax.

kmartens27 commented on content/_data/books/Learning_Continuous_Integration_with_Jenkins.yml:
```diff
@@ -0,0 +1,21 @@
+type: Paid
+image: books/continous-integration-with-Jenkins.jpg
+description: |-
+  Speed up the software delivery process and software productivity using the latest features of Jenkins 
+  
+  . Take advantage of a Continuous Integration and Continuous Delivery solution to speed up productivity and achieve faster software delivery 
+  . See all the new features introduced in Jenkins 2.x, such as Pipeline as code, Multibranch pipeline, Docker Plugin, and more 
+  . Learn to implement Continuous Integration and Continuous Delivery by orchestrating multiple DevOps tools using Jenkins 
+  . Get to know some of the most popular ways to set up Jenkins 
+  . Manage users, projects, and permissions in Jenkins to ensure better security
+  . Leverage the power of plugins in Jenkins 
+  . Learn how to create a CI pipeline using Jenkins Blue Ocean 
+  . Create a distributed build farm using Docker and use it with Jenkins 
+  . Implement CI and CD using Jenkins 
+  . See the difference between CD and Continuous Deployment 
+  . Understand the concepts of CI In Detail. 
```
```suggestion
In this book, you'll discover how to take advantage of a Continuous Integration and Continuous Delivery solution to speed up productivity and achieve faster software delivery.
You'll read about all the new features introduced in Jenkins 2.x, such as Pipeline as code, Multibranch pipeline, the Docker Plugin, and more.
You'll learn how to implement Continuous Integration and Continuous Delivery, by orchestrating multiple DevOps tools using Jenkins.
You'll learn how to manage users, projects, and permissions in Jenkins to ensure better security.
Finally, you'll learn how to leverage the power of plugins in Jenkins!
```
Similar suggestions as above, with making the description shorter and removing some unnecessary sentences/wordings. Updated others to match syntax and single sentence per line formatting.

This single sentence per line formatting will render in-line on the page, so there is no worry about spacing.

kmartens27 commented on content/_data/books/Mastering_Jenkins.yml:
```diff
@@ -0,0 +1,19 @@
+type: Paid
+image: books/Mastering-Jenkins.jpg
+description: |-
+  Configure and extend Jenkins to architect, build, and automate efficient software delivery pipelines 
+  . Configure and horizontally scale a Jenkins installation to support a development organization of any size 
+  . Implement Continuous Integration, Continuous Delivery, and Continuous Deployment solutions in Jenkins 
+  . A step-by-step guide to help you get the most out of the powerful automation orchestration platform that is Jenkins 
+  . Create and manage various types of build jobs, and implement automation tasks to support a software project of any kind 
+  . Get to grips with the automated testing architecture, and scalable automated testing techniques 
+  . Facilitate the delivery of software across the SDLC by creating scalable automated deployment solutions 
+  . Manage scalable automation pipelines in Jenkins using the latest build, test, and deployment strategies 
+  . Implement a scalable master / slave build automation platform, which can support Windows, Mac OSX, and Linux software solutions 
+  . Cover troubleshooting and advanced configuration techniques for Jenkins slave nodes. 
```
```suggestion
Learn how to configure and extend Jenkins to architect, build, and automate efficient software delivery pipelines.
You'll review how to configure and horizontally scale a Jenkins installation, to support a development organization of any size.
This book also describes how to implement Continuous Integration, Continuous Delivery, and Continuous Deployment solutions in Jenkins.
Discover how automated testing architecture and scalable automated testing techniques can be used to take your Jenkins builds to the next level.
You'll read about how to manage scalable automation pipelines in Jenkins using the latest build, test, and deployment strategies.
```
updated formatting/syntax/length as in other suggestions.

I also deleted any mentions/usage of the words "master" and "slave" as they are no longer accepted terminology and have been replaced by "controller" and "agent".

kmartens27 reviewed (changes requested):
Hi @kanhadharmikk, I was able to do a proper review/provide suggestions on how to update the book entries. The formatting and length was updated across the board, and some entries needed to be removed entirely due to non-inclusive word usage ("master" and "slave").

Please let me know if you have any questions or concerns about my suggestions, and I'd be more than happy to help!

kanhadharmikk commented:
Thank you so much @kmartens27 for all your guidance and help really learning a lot. Totally agree with all your suggestions which you have made.

kanhadharmikk commented:
I have committed all your suggestions and as a beginner in open source, some checks were not successfully and I am not getting why it happen so. Can you help me this.

halkeye commented:
according to https://ci.jenkins.io/blue/organizations/jenkins/Websites%2Fjenkins.io/detail/PR-6043/12/pipeline (details link) in the checks screen

[2023-03-02T17:12:16.082Z] An error occurred: (/home/jenkins/workspace/Websites_jenkins.io_PR-6043/content/_data/books/Hands-on_Pipeline_as_YAML_with_Jenkins.yml): could not find expected ':' while scanning a simple key at line 9 column 1

Looks like you have a yaml error?

zbynek commented:
it would be nice to convert the files to AsciiDoc -- it will help you syntax error like this. Maybeyou can check how e.g. the upgrade guides are rendered and do something similar here.

kmartens27 commented:
Thanks @halkeye and @zbynek for your assistance in determining the issue.

I think the .yml files stem from the [original creation of the Books page](https://github.com/jenkins-infra/jenkins.io/pull/5643). I agree that asciidoc would probably allow for better formatting/consistency, and would be happy to update the existing entries to match.

kanhadharmikk commented:
What's the status of my PR? @kmartens27  can you help me with this?

kmartens27 reviewed (approved):
This is terrific, thanks very much @kanhadharmikk!

MarkEWaite commented:
Please use a different branch name than "master".  When you use a branch named "master", you complicate the review process and complicate your own future interactions with the repository.

See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request for more information

kanhadharmikk commented:
Thank you @MarkEWaite for guiding me !! I will read the documentation suggested by you and will make sure to use different branch name and will not make the same mistake further.

Thank you very much to all the mentors @kmartens27, @MarkEWaite, @halkeye, @zbynek for all your guidance, learning and support, which helps me to learn a lot about Open Source and helps to me make my first contribution.
